sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -925,18 +923,20 @@ public boolean start() throws IOException {
       consumer = source.consumerFactoryFn.apply(source.consumerConfig);
       consumer.assign(source.assignedPartitions);
 
-      // seek to consumedOffset + 1 if it is set
       for (PartitionState p : partitionStates) {
-        if (p.consumedOffset >= 0) {
-          LOG.info("{}: resuming {} at {}", name, p.topicPartition, p.consumedOffset + 1);
-          consumer.seek(p.topicPartition, p.consumedOffset + 1);
+        if (p.nextOffset >= 0) {
+          consumer.seek(p.topicPartition, p.nextOffset);
         } else {
-          LOG.info("{}: resuming {} at default offset", name, p.topicPartition);
+          // Set nextOffset to current position, otherwise checkpoint would contain invalid
```

Comment:
Possible wording change:

```java
// nextOffset is currently uninitialized, meaning "start from latest" record. Ask the consumer for that offset
// and start there, and this information will be persisted across checkpoints if this reader is closed.
```

Sanity check: what does `consumer.position` return if there has never been a record published? Hopefully, it returns `0`?